packages/cli/lib/projects.js

```diff
@@ -304,61 +308,142 @@ const uploadProjectFiles = async (
       })
     );
     if (err.error.subCategory === ERROR_TYPES.PROJECT_LOCKED) {
-      logger.log(i18n(`${i18nKey}.logs.projectLockedError`));
+      logger.log(i18n(`${i18nKey}.uploadProjectFiles.projectLockedError`));
     }
     process.exit(EXIT_CODES.ERROR);
   }
 
   return { buildId };
 };
 
+const pollProjectBuildAndDeploy = async (
+  accountId,
+  projectConfig,
+  tempFile,
+  buildId,
+  silenceLogs = false
+) => {
+  const {
+    autoDeployId,
+    isAutoDeployEnabled,
+    deployStatusTaskLocator,
+    status,
+  } = await pollBuildStatus(accountId, projectConfig.name, buildId, null, true);
+  // autoDeployId of 0 indicates a skipped deploy
+  const isDeploying =
+    isAutoDeployEnabled && autoDeployId > 0 && deployStatusTaskLocator;
+
+  if (!silenceLogs) {
+    uiLine();
+  }
+
+  const result = {
+    succeeded: true,
+    buildId,
+    buildSucceeded: true,
+    autodeployEnabled: isAutoDeployEnabled,
+  };
+
+  if (status === 'FAILURE') {
+    result.buildSucceeded = false;
+    result.succeeded = false;
+    return result;
+  } else if (isDeploying) {
+    if (!silenceLogs) {
+      logger.log(
+        i18n(
+          `${i18nKey}.pollProjectBuildAndDeploy.buildSucceededAutomaticallyDeploying`,
+          {
+            accountIdentifier: uiAccountDescription(accountId),
+            buildId,
+          }
+        )
+      );
+    }
+    const { status } = await pollDeployStatus(
+      accountId,
+      projectConfig.name,
+      deployStatusTaskLocator.id,
+      buildId,
+      true
+    );
+    if (status === 'FAILURE') {
+      result.succeeded = false;
+    }
+  }
+
+  try {
+    tempFile.removeCallback();
+    logger.debug(
+      i18n(`${i18nKey}.pollProjectBuildAndDeploy.cleanedUpTempFile`, {
+        path: tempFile.name,
+      })
+    );
+  } catch (e) {
+    logger.error(e);
+  }
+
+  return result;
+};
+
 const handleProjectUpload = async (
   accountId,
   projectConfig,
   projectDir,
   callbackFunc,
   uploadMessage
 ) => {
-  const i18nKey = 'cli.commands.project.subcommands.upload';
   const srcDir = path.resolve(projectDir, projectConfig.srcDir);
 
   const filenames = fs.readdirSync(srcDir);
   if (!filenames || filenames.length === 0) {
     logger.log(
-      i18n(`${i18nKey}.logs.emptySource`, { srcDir: projectConfig.srcDir })
+      i18n(`${i18nKey}.handleProjectUpload.emptySource`, {
+        srcDir: projectConfig.srcDir,
+      })
     );
     process.exit(EXIT_CODES.SUCCESS);
   }
 
   const tempFile = tmp.fileSync({ postfix: '.zip' });
 
   logger.debug(
-    i18n(`${i18nKey}.debug.compressing`, {
+    i18n(`${i18nKey}.handleProjectUpload.compressing`, {
       path: tempFile.name,
     })
   );
 
   const output = fs.createWriteStream(tempFile.name);
   const archive = archiver('zip');
 
-  output.on('close', async function() {
-    logger.debug(
-      i18n(`${i18nKey}.debug.compressed`, {
-        byteCount: archive.pointer(),
-      })
-    );
+  const result = new Promise(resolve =>
```

Comment:
This `handleProjectUpload` function now returns a promise. I wanted to be able to wait until the callbackFunc had returned the necessary data before continuing. This enables us to make decisions based on the successes/failures that can happen in callbackFunc. There's an example of that [here](https://github.com/HubSpot/hubspot-cli/pull/827/files#diff-1989a521c805539a7ae5e60fa3f7b0f79d83106bbb225045cf33f831abf97757R45).